# Part 4 
## 3.11 Homework
### Erdős-Rényi Networks

Consider an Erdős-Rényi network with N = 3,000 nodes, connected to each other with probability $p = 10^{–3}$.
#### What is the expected number of links, $\langle L \rangle$?

Remember the formula (Box 3.11 in book): $$ \langle L \rangle = \frac{pN \cdot (N-1)}{2}$$ 
So if $N = 3000$ and $p = 10^{–3}$, we can then calculate the expected number of links as:


In [3]:
N = 3000
p = 10**-3

L = (N * (N - 1) / 2) * p
print(L)

4498.5


So the expected number of links is:
 $$
 \langle L \rangle = \frac{10^{-3} \cdot 3000 \cdot (3000-1)}{2} = 4498.5
 $$ 

#### In which regime is the network?

According to chapter 3.6 there are 4 distinct regimes:

Subcritical Regime: where $0 < \langle k \rangle < 1$
   
    - In this regime, most nodes are in small, isolated components, and the network consists mainly of disconnected subgraphs with no giant component.

Critical Point: where $ \langle k \rangle = 1 $
     
     - At this point, a giant component just starts to emerge, and the network undergoes a phase transition where a single large connected component appears, but the network is still mostly fragmented.

Supercritical Regime: where $ \langle k \rangle > 1$
    
    - In this regime, a giant component exists, and the majority of nodes are connected within it, but the network is not fully connected. Some small components may still be present.

Connected Regime: where $ \langle k \rangle > ln(N)$
    
    - In this regime, almost all nodes belong to a single giant component, and the network is fully connected with high probability. The network contains few if any isolated nodes or disconnected components.

So to find out what regime the network is in, we need to calculate $\langle k \rangle$:

$$
\langle k \rangle = p(N-1)
$$

In [4]:
k = p * (N - 1)
print(k)

2.999


$$
\langle k \rangle = 10^{-3}(3000-1) = 2.999
$$

Since $\langle k \rangle = 2.999$ the network must be in the supercritical regimes since $2.999 > 1$. It cannot be in subcritical or critical point, nor in connected regime since $ln(3000) = 8,0063675677$

This means a giant component does exist and most nodes are connected within but the network isn't fully connected.

#### Calculate the probability $p_c$ so that the network is at the critical point.

The critical point occurs when $\langle k \rangle = 1$, So we will use the same formula to calculate $\langle k \rangle$ as above but solve for $p_c$:

$$
p_c(N-1) = 1 <=> p_c = \frac{1}{N-1}
$$

for $N = 3000$:

In [5]:
p_c = 1 / (N - 1)
print(p_c)

0.00033344448149383126


$$
p_c = \frac{1}{3000-1} \approx 0.000333
$$

So when $p_c \approx 0.000333$ then the network is at the critical point.

####  Given the linking probability $p = 10^{–3}, calculate the number of nodes $N^{cr} so that the network has only one component.

If the network only should have one component, then it should be a fully connected network i.e. a connected regime - where $ \langle k \rangle > ln(N)$

Again we'll use the same formula:
$$
\langle k \rangle = p(N-1)
$$

We want to find the 